internal/upstreamoidc/upstreamoidc.go

```diff
@@ -236,36 +238,59 @@ func (p *ProviderConfig) tryRevokeRefreshToken(
+	iss := split[0]
+	sub := split[1]
+	if iss == "" || sub == "" {
+		return "", "", errors.New("downstream subject was malformed")
 	}
-	if validated.AccessTokenHash != "" {
-		if err := validated.VerifyAccessToken(tok.AccessToken); err != nil {
+	return split[0], split[1], nil
+}
+
+// ValidateTokenAndMergeWithUserInfo will validate the ID token. It will also merge the claims from the userinfo endpoint response,
+// if the provider offers the userinfo endpoint.
+func (p *ProviderConfig) ValidateTokenAndMergeWithUserInfo(ctx context.Context, tok *oauth2.Token, expectedIDTokenNonce nonce.Nonce, requireIDToken bool) (*oidctypes.Token, error) {
+	var validatedClaims = make(map[string]interface{})
+
+	idTok, hasIDTok := tok.Extra("id_token").(string)
+	var idTokenExpiry time.Time
+	// if we require the id token, make sure we have it.
+	// also, if it exists but wasn't required, still make sure it passes these checks.
+	// nolint:nestif
+	if hasIDTok || requireIDToken {
+		if !hasIDTok {
+			return nil, httperr.New(http.StatusBadRequest, "received response missing ID token")
+		}
+		validated, err := p.Provider.Verifier(&coreosoidc.Config{ClientID: p.GetClientID()}).Verify(coreosoidc.ClientContext(ctx, p.Client), idTok)
+		if err != nil {
 			return nil, httperr.Wrap(http.StatusBadRequest, "received invalid ID token", err)
 		}
-	}
-	if expectedIDTokenNonce != "" {
-		if err := expectedIDTokenNonce.Validate(validated); err != nil {
-			return nil, httperr.Wrap(http.StatusBadRequest, "received ID token with invalid nonce", err)
+		if validated.AccessTokenHash != "" {
+			if err := validated.VerifyAccessToken(tok.AccessToken); err != nil {
+				return nil, httperr.Wrap(http.StatusBadRequest, "received invalid ID token", err)
+			}
 		}
+		if expectedIDTokenNonce != "" {
+			if err := expectedIDTokenNonce.Validate(validated); err != nil {
+				return nil, httperr.Wrap(http.StatusBadRequest, "received ID token with invalid nonce", err)
+			}
+		}
+		if err := validated.Claims(&validatedClaims); err != nil {
+			return nil, httperr.Wrap(http.StatusInternalServerError, "could not unmarshal id token claims", err)
+		}
+		maybeLogClaims("claims from ID token", p.Name, validatedClaims)
+		idTokenExpiry = validated.Expiry // keep track of the id token expiry if we have an id token. Otherwise, it'll just be the zero value.
 	}
+	idTokenSubject, _ := validatedClaims[oidc.IDTokenSubjectClaim].(string)
 
-	var validatedClaims map[string]interface{}
-	if err := validated.Claims(&validatedClaims); err != nil {
-		return nil, httperr.Wrap(http.StatusInternalServerError, "could not unmarshal id token claims", err)
-	}
-	maybeLogClaims("claims from ID token", p.Name, validatedClaims)
-
-	if err := p.maybeFetchUserInfoAndMergeClaims(ctx, tok, validatedClaims); err != nil {
-		return nil, httperr.Wrap(http.StatusInternalServerError, "could not fetch user info claims", err)
+	if len(idTokenSubject) > 0 || !requireIDToken {
```

Comment:
This was just pulled out of the `maybeFetchUserInfoAndMergeClaims` function and modified to allow the no id token case:
```
if len(idTokenSubject) == 0 {
		return nil // defer to existing ID token validation
	}
```
But yes is odd.

Reply:
I believe it will always fail later on (e.g. `getSubjectAndUsernameFromUpstreamIDToken` wouldn't allow you to not have a subject). But it does seem like it should maybe just fail earlier since not having an ID token subject is invalid.